platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultConfigurationCacheDegradationController.kt

```diff
@@ -0,0 +1,74 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl
+
+import org.gradle.api.Task
+import org.gradle.api.internal.ConfigurationCacheDegradationController
+import org.gradle.api.internal.provider.ConfigurationTimeBarrier
+import org.gradle.api.logging.Logging
+import org.gradle.api.provider.Provider
+import org.gradle.execution.plan.TaskNode
+import org.gradle.internal.cc.impl.services.DeferredRootBuildGradle
+import org.gradle.internal.service.scopes.Scope.BuildTree
+import org.gradle.internal.service.scopes.ServiceScope
+import java.util.concurrent.ConcurrentHashMap
+
+@ServiceScope(BuildTree::class)
+internal class DefaultConfigurationCacheDegradationController(
+    private val deferredRootBuildGradle: DeferredRootBuildGradle,
+    private val configurationTimeBarrier: ConfigurationTimeBarrier,
+) : ConfigurationCacheDegradationController {
+
+    private
+    val logger = Logging.getLogger(DefaultConfigurationCacheDegradationController::class.java)
+
+    private val tasksDegradationRequests = ConcurrentHashMap<Task, List<Provider<String>>>()
+    val degradationReasons by lazy(::collectDegradationReasons)
+
+    override fun requireConfigurationCacheDegradation(task: Task, reason: Provider<String>) {
+        if (!configurationTimeBarrier.isAtConfigurationTime) {
+            logger.debug("Configuration cache degradation request of task {} is ignored at execution time", task.path)
+            return
+        }
+        tasksDegradationRequests.compute(task) { _, reasons -> reasons?.plus(reason) ?: listOf(reason) }
+    }
+
+    private fun collectDegradationReasons(): Map<Task, List<String>> {
+        val result = mutableMapOf<Task, List<String>>()
+        if (tasksDegradationRequests.isNotEmpty()) {
+            deferredRootBuildGradle.gradle.taskGraph.visitScheduledNodes { scheduledNodes, _ ->
+                scheduledNodes.filterIsInstance<TaskNode>().map { it.task }.forEach { task ->
+                    val taskDegradationReasons = tasksDegradationRequests[task]
+                        ?.mapNotNull { evaluateDegradationReason(it) }
+                        ?.sorted()
+
+                    if (!taskDegradationReasons.isNullOrEmpty()) {
+                        result[task] = taskDegradationReasons
+                    }
+                }
+            }
+        }
+        return result
+    }
+
+    // Request is represented by the user code, which can throw. We evaluate such requests to valid degradation reasons.
+    private fun evaluateDegradationReason(request: Provider<String>): String? = try {
+        request.orNull
+    } catch (e: Exception) {
+        e.message ?: e::class.qualifiedName
```

Comment:
:question: I wonder if swallowing the exception is the right thing to do. IMO, failure in the user code (or in the Gradle code, as this is an internal API) should be surfaced.